File: internal/hemcore/pathconfinement/pathconfinement.go
package pathconfinement

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type Roots struct {
	HomeDir     string
	ProjectPath string
}

var blockedHomePrefixes = []string{
	".ssh", ".aws", ".gnupg", ".config", ".local", ".npm", ".docker", ".kube",
	".credentials", ".heroku", ".netrc", ".env", ".gitconfig", ".git-credentials",
	".npmrc", ".bash_profile", ".bashrc", ".zshrc", ".profile", ".pgpass", ".gem",
}

func RootsFromPaths(homeDir, projectPath *string) *Roots {
	switch {
	case homeDir == nil && projectPath == nil:
		return nil
	case homeDir != nil && projectPath != nil:
		return &Roots{HomeDir: *homeDir, ProjectPath: *projectPath}
	case homeDir != nil:
		return &Roots{HomeDir: *homeDir, ProjectPath: *homeDir}
	default:
		return &Roots{HomeDir: *projectPath, ProjectPath: *projectPath}
	}
}

func PathHasTraversal(path string) bool {
	clean := filepath.Clean(path)
	parts := strings.Split(clean, string(os.PathSeparator))
	for _, part := range parts {
		if part == ".." {
			return true
		}
	}
	return strings.Contains(path, "..")
}

func IsStrictlyUnder(resolved, root string) bool {
	if resolved == root {
		return true
	}
	prefix := root + string(os.PathSeparator)
	return strings.HasPrefix(resolved, prefix)
}

func ValidateConstrainedWritePath(dest string, roots *Roots) (string, error) {
	if dest == "" {
		return "", fmt.Errorf("empty destination path")
	}
	if PathHasTraversal(dest) {
		return "", fmt.Errorf(`path traversal detected: "%s" contains ".."`, dest)
	}
	if !filepath.IsAbs(dest) {
		return "", fmt.Errorf("destination must be absolute for confinement check: %s", dest)
	}

	resolved := filepath.Clean(dest)
	if !IsStrictlyUnder(resolved, roots.HomeDir) && !IsStrictlyUnder(resolved, roots.ProjectPath) {
		return "", fmt.Errorf(`path "%s" resolves outside home and project directories`, resolved)
	}

	if IsStrictlyUnder(resolved, roots.HomeDir) {
		rel, err := filepath.Rel(roots.HomeDir, resolved)
		if err == nil {
			for _, prefix := range blockedHomePrefixes {
				if rel == prefix || strings.HasPrefix(rel, prefix+string(os.PathSeparator)) ||
					strings.Contains(rel, string(os.PathSeparator)+prefix+string(os.PathSeparator)) {
					return "", fmt.Errorf(`blocked content path prefix: "%s"`, rel)
				}
			}
		}
	}

	return resolved, nil
}

func ValidateHomeRelativeImportSegment(segment string) error {
	if segment == "" {
		return fmt.Errorf("empty home-relative path segment")
	}
	if strings.Contains(segment, "..") {
		return fmt.Errorf(`path traversal detected: "%s" contains ".."`, segment)
	}
	if filepath.IsAbs(segment) {
		return fmt.Errorf(`path traversal detected: "%s" is absolute`, segment)
	}
	return nil
}

File: crates/hem-core/src/path_confinement.rs
use std::path::{Component, Path, PathBuf};

/// Roots that writable restore/bundle paths must stay within.
#[derive(Debug, Clone)]
pub struct ConfinementRoots {
    pub home_dir: PathBuf,
    pub project_path: PathBuf,
}

const BLOCKED_HOME_PREFIXES: &[&str] = &[
    ".ssh",
    ".aws",
    ".gnupg",
    ".config",
    ".local",
    ".npm",
    ".docker",
    ".kube",
    ".credentials",
    ".heroku",
    ".netrc",
    ".env",
    ".gitconfig",
    ".git-credentials",
    ".npmrc",
    ".bash_profile",
    ".bashrc",
    ".zshrc",
    ".profile",
    ".pgpass",
    ".gem",
];

pub fn path_has_traversal(path: &Path) -> bool {
    path.components()
        .any(|component| matches!(component, Component::ParentDir))
}

pub fn is_strictly_under(resolved: &Path, root: &Path) -> bool {
    let root_str = root.to_string_lossy();
    let resolved_str = resolved.to_string_lossy();
    resolved_str == root_str
        || resolved_str.starts_with(&format!(
            "{root_str}{}",
            std::path::MAIN_SEPARATOR
        ))
}

/// Reject traversal, out-of-root absolute paths, and blocked home prefixes.
pub fn validate_constrained_write_path(
    dest: &Path,
    roots: &ConfinementRoots,
) -> Result<PathBuf, String> {
    if dest.as_os_str().is_empty() {
        return Err("Empty destination path".to_string());
    }
    if path_has_traversal(dest) {
        return Err(format!(
            "Path traversal detected: \"{}\" contains \"..\"",
            dest.display()
        ));
    }
    if !dest.is_absolute() {
        return Err(format!(
            "Destination must be absolute for confinement check: {}",
            dest.display()
        ));
    }

    let resolved = dest.to_path_buf();
    if !is_strictly_under(&resolved, &roots.home_dir)
        && !is_strictly_under(&resolved, &roots.project_path)
    {
        return Err(format!(
            "Path \"{}\" resolves outside home and project directories",
            resolved.display()
        ));
    }

    if is_strictly_under(&resolved, &roots.home_dir) {
        if let Ok(relative) = resolved.strip_prefix(&roots.home_dir) {
            let rel = relative.to_string_lossy();
            for prefix in BLOCKED_HOME_PREFIXES {
                if rel.starts_with(prefix)
                    || rel.starts_with(&format!("{prefix}{}", std::path::MAIN_SEPARATOR))
                    || rel.contains(&format!(
                        "{}{prefix}{}",
                        std::path::MAIN_SEPARATOR,
                        std::path::MAIN_SEPARATOR
                    ))
                {
                    return Err(format!("Blocked content path prefix: \"{rel}\""));
                }
            }
        }
    }

    Ok(resolved)
}

/// Validate a home-relative import path segment before joining to home_dir.
pub fn validate_home_relative_import_segment(segment: &str) -> Result<(), String> {
    if segment.is_empty() {
        return Err("Empty home-relative path segment".to_string());
    }
    if segment.contains("..") {
        return Err(format!(
            "Path traversal detected: \"{segment}\" contains \"..\""
        ));
    }
    if Path::new(segment).is_absolute() {
        return Err(format!(
            "Path traversal detected: \"{segment}\" is absolute"
        ));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_traversal_in_dest() {
        let roots = ConfinementRoots {
            home_dir: PathBuf::from("/home/user"),
            project_path: PathBuf::from("/home/user/project"),
        };
        let err = validate_constrained_write_path(
            Path::new("/home/user/../../etc/passwd"),
            &roots,
        )
        .expect_err("traversal");
        assert!(err.contains("traversal"));
    }

    #[test]
    fn allows_paths_under_home() {
        let roots = ConfinementRoots {
            home_dir: PathBuf::from("/home/user"),
            project_path: PathBuf::from("/home/user/project"),
        };
        let ok = validate_constrained_write_path(
            Path::new("/home/user/.codex/config.toml"),
            &roots,
        )
        .expect("allowed");
        assert_eq!(ok, PathBuf::from("/home/user/.codex/config.toml"));
    }
}